src/Transaction.php

```diff
@@ -0,0 +1,650 @@
+        } else {
+            array_unshift($uniqueMetas, new AccountMeta($feePayer, true, true));
+        }
+
+        // Disallow unknown signers
+        foreach ($this->signatures as $signature) {
+            $uniqueIndex = $this->arraySearchAccountMetaForPublicKey($uniqueMetas, $signature);
+            if ($uniqueIndex > -1) {
+                $uniqueMetas[$uniqueIndex]->isSigner = true;
+            } else {
+                throw new InputValidationException("Unknown signer: {$signature->getPublicKey()->toBase58()}");
+            }
+        }
+
+        $numRequiredSignatures = 0;
+        $numReadonlySignedAccounts = 0;
+        $numReadonlyUnsignedAccounts = 0;
+
+        // Split out signing from non-signing keys and count header values
+        /**
+         * @var array<string> $signedKeys
+         */
+        $signedKeys = [];
+        /**
+         * @var array<string> $unsignedKeys
+         */
+        $unsignedKeys = [];
+
+        foreach ($uniqueMetas as $accountMeta) {
+            if ($accountMeta->isSigner) {
+                array_push($signedKeys, $accountMeta->getPublicKey()->toBase58());
+                $numRequiredSignatures++;
+                if (! $accountMeta->isWritable) {
+                    $numReadonlySignedAccounts++;
+                }
+            } else {
+                array_push($unsignedKeys, $accountMeta->getPublicKey()->toBase58());
+                if (! $accountMeta->isWritable) {
+                    $numReadonlyUnsignedAccounts++;
+                }
+            }
+        }
+
+        // Initialize signature array, if needed
+        if (! $this->signatures) {
+            $this->signatures = array_map(function($signedKey) {
+                return new SignaturePubkeyPair(new PublicKey($signedKey), null);
+            }, $signedKeys);
+        }
+
+        $accountKeys = array_merge($signedKeys, $unsignedKeys);
+        /**
+         * @var array<CompiledInstruction> $instructions
+         */
+        $instructions = array_map(function (TransactionInstruction $instruction) use ($accountKeys) {
+            $programIdIndex = array_search($instruction->programId->toBase58(), $accountKeys);
+            $encodedData = $instruction->data;
+            $accounts = array_map(function (AccountMeta $meta) use ($accountKeys) {
+                return array_search($meta->getPublicKey()->toBase58(), $accountKeys);
+            }, $instruction->keys);
+            return new CompiledInstruction(
+                $programIdIndex,
+                $accounts,
+                $encodedData
+            );
+        }, $this->instructions);
+
+        return new Message(
+            new MessageHeader(
+                $numRequiredSignatures,
+                $numReadonlySignedAccounts,
+                $numReadonlyUnsignedAccounts
+            ),
+            $accountKeys,
+            $recentBlockhash,
+            $instructions
+        );
+    }
+
+    /**
```

Comment:
Can we leave behind a todo or an issue marking why this is commented out? Whether it's something we want to implement but haven't, or whatever else?